go.mod

```
@@ -1,6 +1,8 @@
 module github.com/spacemeshos/poet
 
-go 1.21.1
+go 1.21.0
+
+toolchain go1.21.3
```

I think we can go with

```suggestion
go 1.21.3
```

The difference is that any dependency using `poet` with `go 1.21.3` will also be updated to use that dependency automatically by `go get`. While explicitly stating `toolchain` will not do that: see https://go.dev/doc/toolchain#get

---

Setting `go 1.21.3` would force the consumers of poet as a library (go-sm) to build with go 1.21.3, which is not required. It is required to build poet itself with 1.21.3 and this is accomplished with `toolchain go1.21.3`.

---

Consumers of `poet` will still build their code base with go 1.21.3 (or newer) either way:

```
$ git remote -v
origin  git@github.com:spacemeshos/go-spacemesh.git (fetch)
origin  git@github.com:spacemeshos/go-spacemesh.git (push)

$ go version
go version go1.21.1 linux/arm64

$ go get github.com/spacemeshos/poet@4391f3e
go: downloading github.com/spacemeshos/poet v0.9.8-0.20231011101638-4391f3eec4c9
go: upgraded github.com/spacemeshos/poet v0.9.7 => v0.9.8-0.20231011101638-4391f3eec4c9

$ make build
go: downloading go1.21.3 (linux/arm64)
cd cmd/node ; go build -o /workspaces/spacemesh/go-spacemesh/build/go-spacemesh -ldflags "-X main.version=v1.2.0-55-g21064a75 -X main.commit=21064a75ada70b93bde0ec692aea824baa2c723a -X main.branch=5042-refactor-post-grpc-service-dirty" .
# internal/coverage/rtcov
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# internal/unsafeheader
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# internal/godebugs
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# internal/goarch
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# internal/goexperiment
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# internal/goos
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# math/bits
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# internal/itoa
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# internal/cpu
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# internal/race
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# unicode/utf8
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# crypto/internal/alias
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# runtime/internal/syscall
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# unicode
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# sync/atomic
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# unicode/utf16
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# internal/nettrace
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# vendor/golang.org/x/crypto/cryptobyte/asn1
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# crypto/subtle
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# container/list
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# vendor/golang.org/x/crypto/internal/alias
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# encoding
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# crypto/internal/boring/sig
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# google.golang.org/protobuf/internal/flags
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# github.com/anacrolix/chansync/events
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# github.com/google/go-cmp/cmp/internal/flags
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# github.com/libp2p/go-libp2p/core/control
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# golang.org/x/exp/constraints
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# golang.org/x/crypto/internal/alias
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# golang.org/x/net/internal/iana
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# github.com/quic-go/quic-go/internal/utils/ringbuffer
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# go.uber.org/dig/internal/graph
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# go.opencensus.io/internal/tagencoding
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# go.opentelemetry.io/otel/metric/embedded
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# github.com/oasisprotocol/curve25519-voi/internal/disalloweq
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# google.golang.org/grpc/serviceconfig
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# log/internal
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# cmp
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# github.com/oasisprotocol/curve25519-voi/internal/toolchain
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# hash/maphash
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# golang.org/x/exp/maps
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# github.com/spacemeshos/go-spacemesh/log/errcode
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# github.com/spacemeshos/economics/constants
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# github.com/spacemeshos/go-spacemesh/hare/eligibility/config
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# golang.org/x/text/encoding/internal/identifier
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# golang.org/x/text/internal/utf8internal
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# log/slog/internal
compile: version "go1.21.1" does not match go tool version "go1.21.3"
# github.com/seehuhn/mt19937
compile: version "go1.21.1" does not match go tool version "go1.21.3"
make: *** [Makefile:69: go-spacemesh] Error 1
```

---

Now `toolchain` needs to be added to `go-spacemesh` as well, while if `go.mod` in `poet` would specify `go 1.21.3` the go get would already update the `go.mod` file.

---

Then I fail to understand the difference between go and toolchain directives. That article is too convoluted.